src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -505,6 +532,55 @@ RouterSequentialModifyExecScan(CustomScanState *node)
 }
 
 
+/*
+ * TaskListRequires2PC determines whether the given task list requires 2PC
+ * because the tasks provided operates on a reference table or there are multiple
+ * tasks and the commit protocol is 2PC.
+ *
+ * Note that we currently do not generate tasks lists that involves multiple different
+ * tables, thus we only check the first task in the list for reference tables.
+ */
+bool
+TaskListRequires2PC(List *taskList)
+{
+	Task *task = NULL;
+	bool multipleTasks = false;
+	uint64 anchorShardId = INVALID_SHARD_ID;
+
+	if (taskList == NIL)
+	{
+		return false;
+	}
+
+	task = (Task *) linitial(taskList);
+	if (task->replicationModel == REPLICATION_MODEL_2PC)
+	{
+		return true;
+	}
+
+	/*
+	 * Some tasks don't set replicationModel thus we rely on
+	 * the anchorShardId as well replicationModel.
+	 *
+	 * TODO: Do we ever need replicationModel in the Task structure?
```

Comment:
I actually intended to remove `task->replicationModel `. But, (in the future) could we ever need a task where the `anchorShardId ` doesn't belong to a reference table but we need `2PC` for that task?

Reply:
As a general observation: Metadata lookups during execution tend to get us into trouble so it's better to put it in the plan.

I could see 2PC being a replication model for a distributed table.

Reply:
> though not sure that really applies here.

Actually, we'll hit this metadata read on any sequential modification query that are not targeting a reference table. But, given that we call `LoadShardInterval()` in the `ExecuteSingleModifyTask()`, I thought `ReferenceTableShardId ()` wouldn't add any measurable overhead.

I'm ignoring the overhead for `DDL`s, which seems OK.

(We hit `LookupShardCacheEntry()` in both functions, which is the actual lookup)